tools/system_libs.py

```diff
@@ -90,10 +88,38 @@ def run_build_commands(commands):
   logger.info('compiled %d inputs' % len(commands))
 
 
+def objectfile_sort_key(filename):
+  """Sort object files that we pass to llvm-ar."""
+  # In general, we simply use alphabetical order, but we special case certain
+  # object files such they come first.  For example, `vmlock.o` contains the
+  # definition of `__vm_wait`, but `mmap.o` also contains a dummy/weak definition
+  # for use by `mmap.o` when `vmlock.o` is not otherwise included.
+  #
+  # When another object looks for `__vm_wait` we prefer that it always find the
+  # real definition first and not refer to dummy one (which is really intended to
+  # be local to `mmap.o` but due to the fact the weak aliases can't have internal
+  # linkage).
+  #
+  # The reason we care is that once an object file is pulled into certain aspects
+  # of it cannot be undone/removed by the linker.  For example, static constrcutors,
+  # stub library dependencies.
+  #
+  # In the case of `mmap.o`, once it get included by the linker, it pulls in the
+  # the reverse dependencies of the mmap syscall (memalign).  If this add sorting
+  # is not performed we see a slight regressio in test_metadce_minimal_pthreads
+  # due to memalign being included.
+  basename = os.path.basename(filename)
+  if basename in {'vmlock.o'}:
+    return 'AAA_' + basename
+  else:
+    return basename
```

Comment:
So IIUC, `__vm_wait` is defined in `vmlock` properly, and in a weak form by `mmap`. If mmap appears first then the weak form is linked in. It will be overwritten by the non-weak `vmlock` version later, so the code will work, but meanwhile it pulled in all of the `mmap` file, which adds some overhead. Is that right?

Is there no linker option to make them look forward for a non-weak version?

Reply:
Thats exactly right yes.   I confirmed this behaviour using native gcc and clang too.   The first symbol that is found will be used.   Only later when the object file containing the strong definition is included for some other reason will the strong definition overwrite the weak one.

Reply:
Could we avoid this code then by splitting up `mmap.c` into two files? Then `mmap__vm_wait` would just contain the `__vm_wait` for `mmap`, and linking it in wouldn't bring in all of `mmap`?

Reply:
Yes, there might be another way to do it such as that one.   But `mmap.c` is an upstream musl file so splitting it up for this reason seem hacky and harder to maintain over time.    This hack doesn't seem any worse to me.

Also, I think over time we might find other files to add to this list for similar reasons since musl makes quick heavy use of weak linking.

Reply:
Fair enough, sgtm.